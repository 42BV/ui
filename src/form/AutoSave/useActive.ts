import { useState } from 'react';

export function useActive() {
  return useState<string>();
}
