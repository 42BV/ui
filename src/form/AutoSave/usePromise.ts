import { useRef } from 'react';

export function usePromise<T>() {
  return useRef<Promise<T>>();
}
