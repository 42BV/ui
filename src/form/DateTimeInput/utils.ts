import moment from 'moment';
import { InputMask } from '../Input/Input';
import { DateFormat, TimeFormat } from './types';

export function combineFormat(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): string {
  if (dateFormat !== false && timeFormat !== false) {
    return `${dateFormat} ${timeFormat}`;
  } else if (dateFormat !== false) {
    return dateFormat;
  } else if (timeFormat !== false) {
    return timeFormat;
  }

  throw new Error(
    'DateTimeInput: dateFormat and timeFormat cannot both be false'
  );
}

export function formatToMask(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): InputMask {
  const dateMask = dateFormatToMask(dateFormat);

  const timeMask = timeFormatToMask(timeFormat);

  const dateMaskIsDefined = dateMask.length > 0;

  if (dateMaskIsDefined && timeMask.length > 0) {
    return [...dateMask, ' ', ...timeMask];
  } else if (dateMaskIsDefined) {
    return dateMask;
  } else {
    return timeMask;
  }
}

function timeFormatToMask(timeFormat: TimeFormat): InputMask {
  if (timeFormat === false) {
    return [];
  }

  return timeFormat.split('').map((char) => (char === ':' ? ':' : /\d/));
}

function dateFormatToMask(dateFormat: DateFormat): InputMask {
  if (dateFormat === false) {
    return [];
  }

  const separator = extractSeparator(dateFormat);
  return dateFormat
    .split('')
    .map((char) => (char === separator ? separator : /\d/));
}

function extractSeparator(dateFormat: string): string {
  if (dateFormat.includes('-')) {
    return '-';
  }

  if (dateFormat.includes('/')) {
    return '/';
  }

  if (dateFormat.includes('.')) {
    return '.';
  }

  throw new Error('DateTimeInput: cannot extract separator from dateFormat');
}

export function isDate(
  value: string,
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): boolean {
  // _ indicates a masks
  if (value.includes('_')) {
    return false;
  } else {
    const date = moment(
      value.trim(), // value includes an empty char at the back for some reason.
      combineFormat(dateFormat, timeFormat)
    );

    return date.isValid();
  }
}
