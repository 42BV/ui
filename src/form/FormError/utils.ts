import { ValidationError } from '@42.nl/jarb-final-form';
import { MetaError } from '../types';
import {
  Translation,
  getTranslator
} from '../../utilities/translation/translator';

export function keyForError(error: MetaError) {
  if (typeof error === 'string') {
    return error;
  }

  if ((error as ValidationError).type === undefined) {
    return (error as Translation).key;
  }

  return (error as ValidationError).type;
}

export function errorMessage(error: MetaError): string {
  const translator = getTranslator();

  if (typeof error === 'string') {
    return translator({ key: error, fallback: error });
  }

  if ((error as ValidationError).type === undefined) {
    return translator(error as Translation);
  }

  const validationError = error as ValidationError;

  switch (validationError.type) {
    case 'ERROR_REQUIRED':
      return translator({
        key: 'JarbFinalForm.VALIDATION.REQUIRED',
        data: validationError,
        fallback: `${validationError.label} is required`
      });
    case 'ERROR_MINIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MINIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be bigger than ${validationError.reasons.minimumLength} characters`
      });
    case 'ERROR_MAXIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAXIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be smaller than ${validationError.reasons.maximumLength} characters`
      });
    case 'ERROR_MIN_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MIN_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be more than ${validationError.reasons.minValue}`
      });
    case 'ERROR_MAX_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAX_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be less than ${validationError.reasons.maxValue}`
      });
    case 'ERROR_NUMBER':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER',
        data: validationError,
        fallback: `${validationError.label} is not a number`
      });
    case 'ERROR_NUMBER_FRACTION':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER_FRACTION',
        data: validationError,
        fallback: `${validationError.label} is not a number. Number may have ${validationError.reasons.fractionLength} digits behind the comma`
      });

    default:
      return translator({
        key: 'FormError.UNKNOWN_ERROR',
        fallback: 'Something is wrong'
      });
  }
}
