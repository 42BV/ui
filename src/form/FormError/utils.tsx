import React from 'react';
import { ValidationError } from '@42.nl/jarb-final-form';
import { MetaError } from '../types';
import {
  getTranslator,
  Translation
} from '../../utilities/translation/translator';

export function errorMessage(error: MetaError): React.ReactNode {
  const translator = getTranslator();

  // We consider it translated already
  if (typeof error === 'string' || React.isValidElement(error)) {
    return error;
  }

  if ((error as ValidationError).type === undefined) {
    return translator(error as Translation);
  }

  const validationError = error as ValidationError;

  switch (validationError.type) {
    case 'ERROR_REQUIRED':
      return translator({
        key: 'JarbFinalForm.VALIDATION.REQUIRED',
        data: validationError,
        fallback: `${validationError.label} is required`
      });
    case 'ERROR_MINIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MINIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be bigger than ${validationError.reasons.minimumLength} characters`
      });
    case 'ERROR_MAXIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAXIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be smaller than ${validationError.reasons.maximumLength} characters`
      });
    case 'ERROR_MIN_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MIN_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be more than ${validationError.reasons.minValue}`
      });
    case 'ERROR_MAX_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAX_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be less than ${validationError.reasons.maxValue}`
      });
    case 'ERROR_NUMBER':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER',
        data: validationError,
        fallback: `${validationError.label} is not a number`
      });
    case 'ERROR_NUMBER_FRACTION':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER_FRACTION',
        data: validationError,
        fallback: `${validationError.label} is not a number. Number may have ${validationError.reasons.fractionLength} digits behind the comma`
      });

    default:
      return translator({
        key: 'FormError.UNKNOWN_ERROR',
        fallback: 'Something is wrong'
      });
  }
}
