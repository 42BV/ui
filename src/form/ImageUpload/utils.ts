import picaFn from 'pica';
import { ImageUploadCrop } from './ImageUpload';
import { Buffer } from 'buffer';
Buffer.from('anything','base64');

let picaInstance: any;

export function getPicaInstance() {
  if (picaInstance) {
    return picaInstance;
  }

  picaInstance = picaFn();

  return picaInstance;
}

export function dataUrlToFile(dataUrl: string, fileName: string): File {
  const byteString = Buffer.from(dataUrl.split(',')[1], 'base64');

  // separate out the mime component
  const mimeString = dataUrl.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to a typed array
  const bytes = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    bytes[i] = byteString[i];
  }

  return new File([ bytes ], fileName, { type: mimeString });
}

type AvatarEditorConfig = {
  borderRadius: number;
  width: number;
  height: number;
};

export function cropToAvatarEditorConfig(
  crop: ImageUploadCrop
): AvatarEditorConfig {
  if (crop.type === 'rect') {
    const { width, height } = crop;

    return { borderRadius: 0, width, height };
  } else {
    const { size } = crop;

    /* 
      When Crop is 'circle' a ridiculously high border radius is needed 
      so crop.size of different sizes show a perfect circle. Not sure
      why this occurs.
    */
    const borderRadius = 9000000;

    return { borderRadius, width: size, height: size };
  }
}

// Calculate the next 'scale' to use based on how hard the user scrolled
export function calculateScale(scale: number, delta: number): number {
  const zoomFactor = Math.max(Math.abs(delta) / 1000, 0.1);
  const nextScale = delta > 0 ? scale - zoomFactor : scale + zoomFactor;
  return Math.min(10, Math.max(1, nextScale));
}

// Replace any part after the last dot in the filename with .png
export function replaceFileExtension(fileName: string): string {
  return fileName.replace(/\.\w+$/, '.png');
}

export function readFile(file: File, callback: (result: string) => void) {
  const reader = new FileReader();
  reader.onloadend = () => {
    /* istanbul ignore else */
    if (typeof reader.result === 'string') {
      callback(reader.result);
    }
  };
  reader.readAsDataURL(file);
}
