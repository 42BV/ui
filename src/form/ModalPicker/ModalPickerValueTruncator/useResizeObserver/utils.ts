export function setResizeObserver(newResizeObserver: any) {
  // @ts-expect-error Accept that ResizeObserver might exists for this browser
  const oldResizeObserver = window.ResizeObserver
    ? // @ts-expect-error Accept that ResizeObserver might exists for this browser
      window.ResizeObserver
    : undefined;

  // @ts-expect-error Accept that ResizeObserver might exists for this browser
  window.ResizeObserver = newResizeObserver;

  return oldResizeObserver;
}
