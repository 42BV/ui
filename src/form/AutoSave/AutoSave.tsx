import React from 'react';
import { FormSpy } from 'react-final-form';
import { FormState } from 'final-form';
import { isEqual } from 'lodash';

import { useActive } from './useActive';
import { usePromise } from './usePromise';

type Props<FormValues, T> = {
  /**
   * Callback fired on blur when value of the form element changed.
   *
   * @param values
   */
  onSave: (values: FormValues) => Promise<T> | undefined;

  /**
   * The initial values to compare to.
   */
  initialValues?: FormValues;
};

export function AutoSave<FormValues, T>(props: Props<FormValues, T>) {
  const { onSave, initialValues } = props;
  const [ active, setActive ] = useActive();
  const promise = usePromise<T>();

  async function onChange(formState: FormState<FormValues>) {
    if (active === formState.active) {
      return;
    }

    const activeAsString = formState.active as string;
    setActive(activeAsString);

    if (!active) {
      return;
    }

    if (promise.current) {
      await promise.current;
    }

    if (!isEqual(formState.values, initialValues)) {
      promise.current = onSave(formState.values);
    }
  }

  return (
    <FormSpy
      subscription={{ values: true, active: true }}
      onChange={onChange}
    />
  );
}
