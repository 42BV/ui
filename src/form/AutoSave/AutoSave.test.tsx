import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { AutoSave } from './AutoSave';
import * as Active from './useActive';
import * as PendingPromise from './usePromise';
import { resolvablePromise } from '../../test/utils';

describe('Component: AutoSave', () => {
  function setup({
    activeField,
    initialValues
  }: {
    activeField?: string;
    initialValues?: { [id: string]: string };
  }) {
    const resolvable = resolvablePromise();
    const onSaveSpy = jest.fn(() => resolvable.promise);
    const setActiveSpy = jest.fn();
    jest
      .spyOn(Active, 'useActive')
      .mockImplementation(() => [activeField, setActiveSpy]);

    // @ts-ignore
    const autoSave = shallow(
      <AutoSave onSave={onSaveSpy} initialValues={initialValues} />
    );

    return { autoSave, onSaveSpy, setActiveSpy };
  }

  test('ui', () => {
    const autoSave = shallow(<AutoSave onSave={jest.fn()} />);
    expect(toJson(autoSave)).toMatchSnapshot();
  });

  describe('events', () => {
    it('should not save on first time field active', () => {
      const { autoSave, onSaveSpy, setActiveSpy } = setup({});

      // @ts-ignore
      autoSave
        .find('FormSpy')
        .props()
        // @ts-ignore
        .onChange({ active: 'test' });

      expect(setActiveSpy).toBeCalledTimes(1);
      expect(setActiveSpy).toBeCalledWith('test');

      expect(onSaveSpy).toBeCalledTimes(0);
    });

    it('should not save when same field changes', () => {
      const { autoSave, onSaveSpy, setActiveSpy } = setup({
        activeField: 'test'
      });

      // @ts-ignore
      autoSave
        .find('FormSpy')
        .props()
        // @ts-ignore
        .onChange({ active: 'test' });

      expect(setActiveSpy).toBeCalledTimes(0);
      expect(onSaveSpy).toBeCalledTimes(0);
    });

    it('should not save when values do not change', () => {
      const { autoSave, onSaveSpy, setActiveSpy } = setup({
        activeField: 'test',
        initialValues: { test: 'test' }
      });

      // @ts-ignore
      autoSave
        .find('FormSpy')
        .props()
        // @ts-ignore
        .onChange({ active: 'friends', values: { test: 'test' } });

      expect(setActiveSpy).toBeCalledTimes(1);
      expect(setActiveSpy).toBeCalledWith('friends');

      expect(onSaveSpy).toBeCalledTimes(0);
    });

    it('should wait until saving again until the previous save has finished', async done => {
      const { promise, resolve } = resolvablePromise();
      // @ts-ignore
      jest
        .spyOn(PendingPromise, 'usePromise')
        .mockReturnValue({ current: promise });

      const { autoSave, onSaveSpy, setActiveSpy } = setup({
        activeField: 'test',
        initialValues: { test: 'test' }
      });

      // @ts-ignore
      autoSave
        .find('FormSpy')
        .props()
        // @ts-ignore
        .onChange({ active: 'friends', values: { test: 'testing' } });

      expect(setActiveSpy).toBeCalledTimes(1);
      expect(setActiveSpy).toBeCalledWith('friends');

      expect(onSaveSpy).toBeCalledTimes(0);

      try {
        await resolve();

        expect(onSaveSpy).toBeCalledTimes(1);
        expect(onSaveSpy).toBeCalledWith({ test: 'testing' });

        done();
      } catch (e) {
        console.error(e);
        done.fail();
      }
    });

    it('should save when another field changes', () => {
      const { autoSave, onSaveSpy, setActiveSpy } = setup({
        activeField: 'test'
      });

      // @ts-ignore
      autoSave
        .find('FormSpy')
        .props()
        // @ts-ignore
        .onChange({ active: 'friends', values: { test: 'test' } });

      expect(setActiveSpy).toBeCalledTimes(1);
      expect(setActiveSpy).toBeCalledWith('friends');

      expect(onSaveSpy).toBeCalledTimes(1);
      expect(onSaveSpy).toBeCalledWith({ test: 'test' });
    });
  });
});
