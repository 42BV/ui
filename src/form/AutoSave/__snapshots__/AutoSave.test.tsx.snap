// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AutoSave ui 1`] = `
<FormSpy
  onChange={[Function]}
  subscription={
    Object {
      "active": true,
      "values": true,
    }
  }
/>
`;
